### PR TITLE
[BugFix] fix ernie-ctm ci bug

### PR DESCRIPTION
**Pull request by wj-Mcat**
### PR types

Bug fixes

### PR changes

Models

### Description

更新 ctm 中不兼容的部分接口

### CODE REVIEW

File: examples/text_to_knowledge/ernie-ctm/train.py [hunk 1]
```diff
@@ -157,7 +158,8 @@ def do_train(args):
             global_step += 1
             input_ids, token_type_ids, seq_len, tags = batch
 
-            loss, _ = model(input_ids, token_type_ids, lengths=seq_len, tag_labels=tags)
+            loss = model(input_ids, token_type_ids, lengths=seq_len, tag_labels=tags)[0]
+
```
Comment (wj-Mcat) [inline examples/text_to_knowledge/ernie-ctm/train.py:164]: change is here

File: paddlenlp/transformers/ernie_ctm/configuration.py [hunk 2]
```diff
@@ -101,6 +98,7 @@ class ErnieCtmConfig(PretrainedConfig):
     """
     model_type = "ernie-ctm"
     pretrained_init_configuration = ERNIE_CTM_PRETRAINED_INIT_CONFIGURATION
+    attribute_map = {"num_tag": "num_labels"}
```
Comment (wj-Mcat) [inline paddlenlp/transformers/ernie_ctm/configuration.py]: change is here

**Comment by wj-Mcat** [comment]
在最新模型当中使用nn.TransformerLayer来构建，权重有所变动。

目前 bos 上面 ctm 模型的权重名称都调整完了，使用以下代码测试是不会出现mismatched keys:

```python
    from paddlenlp.transformers import ErnieCtmForTokenClassification, ErnieCtmNptagModel, ErnieCtmWordtagModel
    model = ErnieCtmWordtagModel.from_pretrained("ernie-ctm")
```

File: paddlenlp/transformers/ernie_ctm/modeling.py [hunk 3]
```diff
@@ -705,12 +711,14 @@ class ErnieCtmForTokenClassification(ErnieCtmPretrainedModel):
...
 
-    def __init__(self, ernie_ctm, num_classes=2, dropout=None):
-        super(ErnieCtmForTokenClassification, self).__init__()
-        self.num_classes = num_classes
-        self.ernie_ctm = ernie_ctm  # allow ernie_ctm to be config
-        self.dropout = nn.Dropout(dropout if dropout is not None else self.ernie_ctm.config["hidden_dropout_prob"])
-        self.classifier = nn.Linear(self.ernie_ctm.config["hidden_size"], num_classes)
+    def __init__(self, config: ErnieCtmConfig):
+        super(ErnieCtmForTokenClassification, self).__init__(config)
+        self.num_classes = config.num_labels
```
Comment (sijunhe) [inline paddlenlp/transformers/ernie_ctm/modeling.py]: 这里是num_classes, 上面的attribute map里面是num_tags， 有没有问题？
num_classes, num_labels, num_tags 这三个做一下取舍

File: paddlenlp/transformers/ernie_ctm/modeling.py [hunk 4]
```diff
@@ -744,7 +752,7 @@ def forward(self, input_ids, token_type_ids=None, position_ids=None, attention_m
 
         """
 
-        sequence_output, _, _ = self.ernie_ctm(
+        sequence_output, _, _, _ = self.ernie_ctm(
```
Comment (sijunhe) [inline paddlenlp/transformers/ernie_ctm/modeling.py]: 这里太丑，风格还是和之前做到一致吧，类似这种
```
        outputs = self.ernie_ctm(
            input_ids=input_ids,
            token_type_ids=token_type_ids,
            attention_mask=attention_mask,
            position_ids=position_ids,
            output_hidden_states=output_hidden_states,
            output_attentions=output_attentions,
            return_dict=return_dict,
        )
```

File: paddlenlp/transformers/ernie_ctm/modeling.py [hunk 5]
```diff
@@ -705,12 +711,14 @@ class ErnieCtmForTokenClassification(ErnieCtmPretrainedModel):
...
+        super(ErnieCtmForTokenClassification, self).__init__(config)
+        self.num_classes = config.num_labels
+        self.ernie_ctm = ErnieCtmModel(config)
+        self.dropout = nn.Dropout(
+            config.classifier_dropout if config.classifier_dropout is not None else config.hidden_dropout_prob
+        )
+        self.classifier = nn.Linear(config.hidden_size, config.num_labels)
         self.apply(self.init_weights)
 
     def forward(self, input_ids, token_type_ids=None, position_ids=None, attention_mask=None):
```
Comment (sijunhe) [inline paddlenlp/transformers/ernie_ctm/modeling.py:724]: 顺便配一下输入参数，和之前一致

File: examples/text_to_knowledge/ernie-ctm/train.py [hunk 6]
```diff
@@ -70,7 +73,7 @@ def evaluate(model, metric, data_loader, tags, tags_to_idx):
     losses = []
     for batch in data_loader():
         input_ids, token_type_ids, seq_len, tags = batch
-        loss, seq_logits = model(input_ids, token_type_ids, lengths=seq_len, tag_labels=tags)
+        loss, seq_logits = model(input_ids, token_type_ids, lengths=seq_len, tag_labels=tags)[:2]
```
Comment (sijunhe) [inline examples/text_to_knowledge/ernie-ctm/train.py:76]: 要不然我们直接用return_dict吧，这种`[:2]`比较迷，不如直接返回一个outputs, 然后outputs.loss, outputs.logits

File: examples/text_to_knowledge/ernie-ctm/train.py [hunk 7]
```diff
@@ -157,7 +158,8 @@ def do_train(args):
             global_step += 1
             input_ids, token_type_ids, seq_len, tags = batch
 
-            loss, _ = model(input_ids, token_type_ids, lengths=seq_len, tag_labels=tags)
+            loss = model(input_ids, token_type_ids, lengths=seq_len, tag_labels=tags)[0]
```
Comment (sijunhe) [inline examples/text_to_knowledge/ernie-ctm/train.py:163]: 同上

File: examples/text_to_knowledge/ernie-ctm/train.py [hunk 8]
```diff
@@ -104,7 +107,7 @@ def do_train(args):
 
     trans_func = partial(convert_example, tokenizer=tokenizer, max_seq_len=args.max_seq_len, tags_to_idx=tags_to_idx)
 
-    batchify_fn = lambda samples, fn=Tuple(
+    batchify_fn = lambda samples, fn=Tuple(  # noqa: E731
```
Comment (sijunhe) [inline examples/text_to_knowledge/ernie-ctm/train.py]: 直接改成这样吧 https://github.com/PaddlePaddle/PaddleNLP/blob/ee70111bfbade68b1120628aa73d19aeed080f31/examples/information_extraction/waybill_ie/run_ernie.py#L120

File: paddlenlp/transformers/ernie_ctm/modeling.py (hunk 3 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/ernie_ctm/modeling.py]: 关于这个问题我和原始作者之前沟通过，他坚持使用：`num_tags`，理由是保持模型参数命名与模型的语义适配性和特殊性。
所以就没有调整。

File: paddlenlp/transformers/ernie_ctm/modeling.py (hunk 4 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/ernie_ctm/modeling.py]: ok

File: paddlenlp/transformers/ernie_ctm/modeling.py (hunk 3 above)
Comment (sijunhe) [reply paddlenlp/transformers/ernie_ctm/modeling.py]: num_tags没问题，那这里是不是应该`self.num_labels = config.num_labels`, 然后如果用户model.num_tags的话attribute_map可以解决

File: paddlenlp/transformers/ernie_ctm/modeling.py (hunk 5 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/ernie_ctm/modeling.py:724]: done

File: examples/text_to_knowledge/ernie-ctm/train.py (hunk 6 above)
Comment (wj-Mcat) [reply examples/text_to_knowledge/ernie-ctm/train.py:76]: 还是算了吧，这个如果要调整的话这个模块的其它模型就都要调整，这个阶段下这个模型可以暂时不调整吧。

File: examples/text_to_knowledge/ernie-ctm/train.py (hunk 8 above)
Comment (wj-Mcat) [reply examples/text_to_knowledge/ernie-ctm/train.py]: done

**Comment by wj-Mcat** [comment]
我在分析原始wordtag模型的时候发现其中有`sent_classifier.weight`和`sent_classifier.bias`权重:

https://github.com/PaddlePaddle/PaddleNLP/blob/21714c3797149f5283c9229313cb93fcb2c2d51a/paddlenlp/taskflow/knowledge_mining.py#L178-L205

可是我在最新的代码中已经找不到其相关代码了，经过分析可知：其删除是在https://github.com/PaddlePaddle/PaddleNLP/pull/1329 中做的调整。可是我不知道任何关于为何要删除此权重的说明。

@linjieccc 不知道你还能回忆其为何要删除呢？

**Comment by wj-Mcat** [comment]
我发现`ernie_ctm/configuration`下的配置和 `taskflow/knowledge_mining`中的配置权重是不一致的，而且我关注到后者中有一个 version 的 tag：`knowledge_mining/wordtag_v1.3/model_state.pdparams`。

所以目前 wordtag 的最高版本是 1.3 吗？`ernie_ctm/configuration`中的链接配置也应该是需要更新的吗？ @linjieccc

**Comment by linjieccc** [comment]
> 我发现`ernie_ctm/configuration`下的配置和 `taskflow/knowledge_mining`中的配置权重是不一致的，而且我关注到后者中有一个 version 的 tag：`knowledge_mining/wordtag_v1.3/model_state.pdparams`。
> 
> 所以目前 wordtag 的最高版本是 1.3 吗？`ernie_ctm/configuration`中的链接配置也应该是需要更新的吗？ @linjieccc

ernie-ctm目前把TranformerEncoder变成LayerList[TransformerEncoderLayer]了，因为相关的模型权重需要对应更新，旧的模型权重没办法兼容新的组网 #3945

**Comment by wj-Mcat** [comment]
是的，目前这个 PR 就是在做这类的兼容升级，其中任务也包含权重文件的升级。

可是目前wordtag的权重在两处有保存：

* https://bj.bcebos.com/paddlenlp/taskflow/knowledge_mining/wordtag_v1.3/model_state.pdparams
* https://bj.bcebos.com/paddlenlp/models/transformers/ernie_ctm/wordtag_pos.pdparams

所以，@linjieccc 这两处的权重是否要统一一下呢？

**Comment by wj-Mcat** [comment]
最近出现了ner和word_segmentation的相关问题，究其原因都是由于之前 ernie_ctm 的更新导致的：
 
1.    没有把num_tag信息添加到配置当中去，于是 crf.transitions 权重没有正确初始化，所以就会分词、分割不正确
    a.    需要更新权重文件
    b.    需要更新配置相关文件
    c.    需要统一 wordtag 最新版本文件
2.  针对于 wordtag 模型目前存在两套模型：word_pos 和 v1.3/model_state.pdparams，其中 crf.transitions 权重shape 不一样，所以就导致了crf.transitions 权重random初始化

以上关于权重和 taskflow 相关的问题，已和 @linjieccc 沟通，他目前正在修复，所以此 pr 需要等待他的完成才可以继续推进。

File: paddlenlp/transformers/ernie_ctm/modeling.py [hunk 9]
```diff
@@ -744,10 +752,11 @@ def forward(self, input_ids, token_type_ids=None, position_ids=None, attention_m
 
         """
 
-        sequence_output, _, _ = self.ernie_ctm(
+        output = self.ernie_ctm(
             input_ids, token_type_ids=token_type_ids, position_ids=position_ids, attention_mask=attention_mask
         )
 
+        sequence_output = output[0]
```
Comment (wawltor) [inline paddlenlp/transformers/ernie_ctm/modeling.py:759]: 这里的改动需要check一下低版本paddle版本是否支持动转静，`output[0]` 这个slice操作是在最新的paddle版本才支持的

Comment (wj-Mcat) [reply paddlenlp/transformers/ernie_ctm/modeling.py:759]: ## 环境

```shell
paddle-bfloat              0.1.7
paddle-cppt                0.0.1
paddle2onnx                0.9.8
paddlefsl                  1.1.0
paddlehub                  2.1.0
paddleocr                  2.5.0.3
paddlepaddle               2.3.2
```
## check 代码

```python
import paddle
from paddlenlp.transformers import ErnieCtmForTokenClassification

model = ErnieCtmForTokenClassification.from_pretrained("wordtag")

inputs = [paddle.static.InputSpec(shape=[None, None], dtype="int64")]

# Convert to static graph with specific input description
model = paddle.jit.to_static(model, input_spec=inputs)
# Save in static graph model.
paddle.jit.save(model, "./pretrained/ctm_static")

static_model = paddle.jit.load("./pretrained/ctm_static")
result = static_model(
    paddle.randint(30, 300, shape=[1, 40])
)
print(result)
```

在以上环境和测试代码中是能 work 的。  @wawltor 在这个条件下是否可以合入呢？

**Review by wawltor — APPROVED** [review]
LGTM
